### ISSUE TITLE
[BUG][Return-types] `Counts` does not work with `default.mixed`

### ISSUE THREAD

**Issue by rmoyard** [PennyLaneAI/pennylane#2984]
### Expected behavior

With `qml.enable_return` activated, `Counts` should work with interfaces (torch, tf, jax) and `default.mixed` but it is not the case.

### Actual behavior

```
import pennylane as qml
import torch

shots=100
qml.enable_return()

measurements = [qml.counts(qml.PauliZ(0)), qml.counts(wires=[0])]

devices = ["default.mixed"]


for m in measurements:
    for d in devices:
        dev = qml.device(d, wires=2, shots=shots)

        def circuit(x):
            qml.Hadamard(wires=[0])
            qml.CRX(x, wires=[0, 1])
            return qml.apply(m)

        qnode = qml.QNode(circuit, dev, diff_method=None)
        res = qnode(torch.tensor(0.5, requires_grad=True))
        print(res)
```

### Tracebacks

```shell
Traceback (most recent call last):
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/testp/test.py", line 24, in <module>
    res = qnode(torch.tensor(0.5, requires_grad=True))
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/qnode.py", line 625, in __call__
    res = qml.execute_new(
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/interfaces/execution.py", line 603, in execute_new
    qml.interfaces.cache_execute(
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/interfaces/execution.py", line 197, in wrapper
    res = fn(execution_tapes.values(), **kwargs)
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/interfaces/execution.py", line 122, in fn
    return original_fn(tapes, **kwargs)
  File "/usr/local/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/_qubit_device.py", line 614, in batch_execute_new
    res = self.execute_new(circuit)
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/_qubit_device.py", line 402, in execute_new
    self._samples = self.generate_samples()
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/_qubit_device.py", line 1023, in generate_samples
    samples = self.sample_basis_states(number_of_states, rotated_prob)
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/pennylane/pennylane/_qubit_device.py", line 1053, in sample_basis_states
    return np.random.choice(basis_states, shots, p=state_probability)
  File "mtrand.pyx", line 919, in numpy.random.mtrand.RandomState.choice
  File "/Users/romainmoyard/PennyLaneMaster/PennyLane/penv/lib/python3.9/site-packages/torch/_tensor.py", line 734, in __array__
    return self.numpy().astype(dtype, copy=False)
RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.
```

### System information

```shell
-
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by timmysilv** [comment]
Note that this is also true for `qml.sample`. The exact reason: the new system [doesn't expand tapes](https://github.com/PennyLaneAI/pennylane/blob/351c6bdb4cffb52609e46245b8fd7c701dc28cd4/pennylane/interfaces/execution.py#L363) for `default.mixed` because that device should support all interfaces. The old system doesn't have that `if default.qubit` check, so it would unwrap all parameters, use numpy, and `np.random.choice` wouldn't freak out about the probabilities being non-numpy

**Comment by obliviateandsurrender** [comment]
This should now be fixed with #5514.

**Comment by PietropaoloFrisoni** [comment]
@obliviateandsurrender  Actually, I cannot run the above code anymore since `qml.enable_return` has been removed from pennylane in release 0.33.
